libcaliptra/src/caliptra_api.c

```diff
@@ -0,0 +1,211 @@
+
+    return 0;
+}
+
+/**
+ * caliptra_mailbox_read_buffer
+ *
+ * Read a mailbxo FIFO into a buffer
+ *
+ * @param[in] buffer A pointer to a valid caliptra_buffer struct
+ *
+ * @return int 0 if successful, -EINVAL if the buffer is too small or the buffer pointer is invalid.
+ */
+static int caliptra_mailbox_read_buffer(struct caliptra_buffer *buffer)
+{
+
+    // Check we have enough room in the buffer
+    if (buffer->len < caliptra_mbox_read_dlen() || !buffer->data)
+       return -EINVAL;
+
+    uint32_t remaining_len = caliptra_mbox_read_dlen();
+    uint32_t *data_dw = (uint32_t *)buffer->data;
+
+    // Copy DWord multiples
+    while (remaining_len > sizeof(uint32_t)) {
+        *data_dw++ = caliptra_mbox_read(MBOX_CSR_MBOX_DATAOUT);
+        remaining_len -= sizeof(uint32_t);
+    }
+
+    // if un-aligned dword reminder...
+    if (remaining_len) {
+        uint32_t data = caliptra_mbox_read(MBOX_CSR_MBOX_DATAOUT);
+        memcpy(data_dw, &data, remaining_len);
+    }
+
+    return 0;
+}
+
+/**
+ * caliptra_mailbox_execute
+ *
+ * Execute a mailbox command and send/retrieve a buffer
+ *
+ * @param[in] cmd Caliptra command opcode
+ * @param[in] mbox_tx_buffer Transmit buffer
+ * @param[in] mbox_rx_buffer Receive buffer
+ *
+ * @return 0 if successful, -EBUSY if the mailbox is locked, -EIO if the command has failed or data is not available or the FSM is not include
+ */
+int caliptra_mailbox_execute(uint32_t cmd, struct caliptra_buffer *mbox_tx_buffer, struct caliptra_buffer *mbox_rx_buffer)
+{
+    // If mbox already locked return
+    if (caliptra_mbox_is_lock()) {
+        return -EBUSY;
+    }
+                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 
+    // Write Cmd and Tx Buffer
+    caliptra_mbox_write_cmd(cmd);
+    caliptra_mailbox_write_fifo(mbox_tx_buffer);
+
+    // Set Execute bit and wait
+    caliptra_mbox_write_execute_wait(true);
+
+    // Check the Mailbox Status
+    uint32_t status = caliptra_mbox_read_status();
+    if (status == CALIPTRA_MBOX_STATUS_CMD_FAILURE) {
+        caliptra_mbox_write_execute(false);
+        return -EIO;
+    } else if(status == CALIPTRA_MBOX_STATUS_CMD_COMPLETE) {
+        caliptra_mbox_write_execute(false);
+        return 0;
+    } else if (status != CALIPTRA_MBOX_STATUS_DATA_READY) {
+        return -EIO;
+    }
+
+    // Read Mbox out Data Len
+    uint32_t dlen = caliptra_mbox_read_dlen();
```

Comment:
Are we discarding dlen?

Reply:
We are. The same thing happens in the c-binding example API. We don't have any real build flags to require we do something with it, so then the question is if reading dlen has any side-effects within the caliptra hardware/rom. If not then we can remove this.

Reply:
Reading dlen does not change state.

Reply:
Ok. Then I'll double check and remove it.